designs/0100-reduce_sum.md

```diff
@@ -0,0 +1,169 @@
+- Feature Name: Reduce Sum Parallelization
+- Start Date: 2020-03-10
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+The reduce_sum function is an attempt to add a parallelization utility that is much easier to use than map-rect without requiring a substantial changes to the autodiff stack.
+
+# Motivation
+[motivation]: #motivation
+
+The overall goal of reduce_sum is to make it easier for users to parallelize their models by streamlining how arguments are handled, hiding work scheduling, and making it more difficult to program something that will accidentally be inefficient.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+reduce_sum is effectively a parallel-for combined with a sum.
+
+For a set of input arguments, ```args0, args1, args2, ...``` and a scalar function ```f```, reduce_sum computes the scalar:
+
+```f(args0) + f(args1) + f(args2) + ...```
+
+In terms of probabilistic models, this is useful when there are N terms in a likelihood that combine multiplicatively. In this case, computing the log likelihood means computing the sum of the log of each of these terms. If a function ```f``` can be written that computes an individual log likelihood, then reduce_sum can be used to compute them all in parallel and accumulate the results.
+
+The actual meaning of N is flexible here. In a regression model, N might be the N different rows of the input dataframe, each which can be evaluated separately. In a hierarchical model, N might correspond to different groups -- perhaps there is some complex shared calculation for each group but these can all be done in parallel.
+
+reduce_sum is not useful, however, when there are dependencies between these likelihood terms. This can happen, for instance, if there were N terms in a Gaussian process likelihood. reduce_sum will not be useful for accelerating this.
+
+The signature we settled on for reduce_sum (which balances the different tradeoffs described in the Summary) is:
+
+```
+real reduce_sum(F func, T[] sliced_arg, int grainsize, T1 arg1, T2 arg2, ...)
+```
+
+where the function ```func``` has a signature like:
+
+```
+real func(int start, int end, T[] sliced_arg, T1 arg1, T2 arg2, ...)
+```
+
+The array ```sliced_arg``` in each case represents the list of calculations to perform and sum. The user-provided function ```func``` is expect to compute the ```start``` through ```end``` terms of the overall sum, accumulate them, and return that value. The trailing arguments ```argN``` are passed without modification to every call of ```func```.
```

Comment:
Should we also provide a user function to slice matrices and vectors?

Reply:
I'm for arrays of things only (now that I look at this I'm not sure we support ```T[]``` in the current implementation -- may only be ```real[]```). I don't think we should slice vectors and matrices. I forget why I think this, something about language consistency, but I settled on this last time we talked through this.

Reply:
I guess the last conversation is roughly between: https://github.com/stan-dev/math/pull/1616#issuecomment-574861041 and https://github.com/stan-dev/math/pull/1616#issuecomment-575671089

Reply:
Yes, that's what I recall from the last discussion we had on this. For now we only allow to slice arrays of T (T being whatever), but nothing else. Reasons were language consistency, memory locality in-efficiency for (column major) Eigen matrices and probably more things I forgot.

Reply:
Right now the implementation in stan math can accept both eigen vectors and std::vectors. I put that in so that it would be easier for brms etc. to use this.

idk, personally I'm fine with not accepting matrices since that feels like a whole bag of Qs on how to do that. Though it feels like taking in a vector would be nice

Reply:
I agree in that slicing a vector would be nice, but I think we ended up saying that a `to_array_1d` is not too expensive to do. From my memory we did decide this matter for the moment, so in the interest of moving forward we should stick with that.

Reply:
Bob said `to_array` is rather expensive

https://github.com/stan-dev/math/pull/1616#issuecomment-575671571

Reply:
`to_array` does an alloc and a copy.  That's just memory pressure and time as there's no autodiff implicated.